rust/data_structures/disjoint_set.rs

```diff
@@ -0,0 +1,116 @@
+/// A Disjoint-set data structure (also called a union–find data structure or merge–find set) 
+/// is a data structure that tracks a set of elements partitioned into a number of 
+/// disjoint (non-overlapping) subsets. It provides near-constant-time operations 
+/// (bounded by the inverse Ackermann function) to add new sets, to merge existing sets,
+/// and to determine whether elements are in the same set.
+/// - https://en.wikipedia.org/wiki/Disjoint-set_data_structure
+use std::cmp::Ordering;
+
+#[derive(Debug)]
+struct Node {
+    id: usize,
+    parent_id: usize,
+    rank: usize,
+}
+
+impl Node {
+    /// Returns a new Node with itself as parent and rank zero
+    fn new(id: usize) -> Node {
+        Node {
+            id: id,
+            parent_id: id,
+            rank: 0,
+        }
+    }
+}
+
+#[derive(Debug)]
+struct DisjointSet {
+    elems: Vec<Node>
+}
+
+impl DisjointSet {
+    /// Create a new DisjointSet with fixed number of elements
+    /// 0 to size - 1, each belonging to a seperate subset
+    fn new(size: usize) -> DisjointSet {
+        DisjointSet {
+            elems: (0..size).map(Node::new).collect()
+        }
+    }
+
+    /// Return the root id of the element with given id,
+    /// also known as the representative node of that particular set
+    fn find(&mut self, id: usize) -> usize {
```

Comment:
If `id` does not exist in the Disjoint Set (which should be the case only when `id >= size`), the program will probably panic in `fn parent` when `self.elems[id]` tries to access something out of bounds.
Not sure if you want to handle/prevent that.

Reply:
Yeah, true. basically every method will panic if an non-existing id is passed. I guess you could handle it by returning `Option` instead, but it might obfuscate the actual data structure logic for this purpose